File: Raven.Abstractions/Extensions/MetadataExtensions.cs
//-----------------------------------------------------------------------
// <copyright file="MetadataExtensions.cs" company="Hibernating Rhinos LTD">
//     Copyright (c) Hibernating Rhinos LTD. All rights reserved.
// </copyright>
//-----------------------------------------------------------------------
using System;
using System.Collections.Generic;
using System.Collections.Specialized;
using System.Globalization;
using System.Linq;
using System.Text;
using Raven.Abstractions.Data;
using Raven.Imports.Newtonsoft.Json;
using Raven.Json.Linq;
using System.Net.Http;

namespace Raven.Abstractions.Extensions
{
    /// <summary>
    /// Extensions for handling metadata
    /// </summary>
    public static class MetadataExtensions
    {
        private static readonly HashSet<string> HeadersToIgnoreClient = new HashSet<string>(StringComparer.OrdinalIgnoreCase)
        {
            // Raven internal headers
            Constants.RavenServerBuild,
            "Raven-Client-Version",
            "Non-Authoritative-Information",
            "Raven-Timer-Request",
            "Raven-Authenticated-User",
            Constants.RavenLastModified,
            "Has-Api-Key",

            // COTS
            "Access-Control-Allow-Origin",
            "Access-Control-Max-Age",
            "Access-Control-Allow-Methods",
            "Access-Control-Request-Headers",
            "Access-Control-Allow-Headers",

            //proxy
            "Reverse-Via",
            "Persistent-Auth",
            "Allow",
            "Content-Disposition",
            "Content-Encoding",
            "Content-Language",
            "Content-Location",
            "Content-MD5",
            "Content-Range",
            "Content-Type",
            "Expires",
            // ignoring this header, we handle this internally
            Constants.LastModified,
            // Ignoring this header, since it may
            // very well change due to things like encoding,
            // adding metadata, etc
            "Content-Length",
            // Special things to ignore
            "Keep-Alive",
            "X-Powered-By",
            "X-AspNet-Version",
            "X-Requested-With",
            "X-SourceFiles",
            // Request headers
            "Accept-Charset",
            "Accept-Encoding",
            "Accept",
            "Accept-Language",
            "Authorization",
            "Cookie",
            "Expect",
            "From",
            "Host",
            "If-Match",
            "If-Modified-Since",
            "If-None-Match",
            "If-Range",
            "If-Unmodified-Since",
            "Max-Forwards",
            "Referer",
            "TE",
            "User-Agent",
            "DNT",
            //Response headers
            "Accept-Ranges",
            "Age",
            "Allow",
            Constants.MetadataEtagField,
            "Location",
            "Retry-After",
            "Server",
            "Set-Cookie2",
            "Set-Cookie",
            "Vary",
            "Www-Authenticate",
            // General
            "Cache-Control",
            "Connection",
            "Date",
            "Pragma",
            "Trailer",
            "Transfer-Encoding",
            "Upgrade",
            "Via",
            "Warning",

            // IIS Application Request Routing Module
            "X-ARR-LOG-ID",
            "X-ARR-SSL",
            "X-Forwarded-For",
            "X-Original-URL",

            // Azure specific
            "X-LiveUpgrade",
            "DISGUISED-HOST",
            "X-SITE-DEPLOYMENT-ID",
        };

        private static readonly HashSet<string> HeadersToIgnoreForClient = new HashSet<string>(HeadersToIgnoreClient.Except(new[]
        {
            Constants.LastModified,
            Constants.RavenLastModified
        }), StringComparer.OrdinalIgnoreCase);

        private static readonly HashSet<string> PrefixesInHeadersToIgnoreClient = new HashSet<string>(StringComparer.OrdinalIgnoreCase)
                                                                               {
                                                                                   "Temp",
                                                                                   "X-NewRelic"
                                                                               };

        /// <summary>
        /// Filters the headers from unwanted headers
        /// </summary>
        /// <param name="self">HttpHeaders to filter</param>
        /// <param name="headersToIgnore">Headers to ignore</param>
        /// <param name="prefixesInHeadersToIgnore">Header prefixes to ignore</param>
        /// <returns></returns>
        public static RavenJObject FilterHeadersToObject(this RavenJObject self, HashSet<string> headersToIgnore, HashSet<string> prefixesInHeadersToIgnore)
        {
            if (self == null)
                return null;

            var metadata = new RavenJObject(self.Comparer);
            foreach (var header in self)
            {
                if (prefixesInHeadersToIgnore.Any(prefix => header.Key.StartsWith(prefix, StringComparison.OrdinalIgnoreCase)))
                    continue;
                if (header.Key == Constants.DocumentIdFieldName)
                    continue;
                if (headersToIgnore.Contains(header.Key))
                    continue;
                var headerName = CaptureHeaderName(header.Key);
                metadata[headerName] = header.Value;
            }
            return metadata;
        }

        /// <summary>
        /// Filters the headers from unwanted headers
        /// </summary>
        /// <param name="self">The self.</param>
        /// <returns></returns>public static RavenJObject FilterHeadersToObject(this System.Collections.Specialized.NameValueCollection self, bool isServerDocument)
        public static RavenJObject FilterHeadersToObject(this RavenJObject self)
        {
            return FilterHeadersToObject(self, HeadersToIgnoreForClient, PrefixesInHeadersToIgnoreClient);
        }

        [Obsolete("Use RavenFS instead.")]
        public static RavenJObject FilterHeadersAttachment(this NameValueCollection self)
        {
            var filterHeaders = self.FilterHeadersToObject();
            if (self["Content-Type"] != null)
                filterHeaders["Content-Type"] = self["Content-Type"];
            return filterHeaders;
        }

        /// <summary>
        /// Filters the headers from unwanted headers
        /// </summary>
        /// <param name="self">The self.</param>
        /// <returns></returns>public static RavenJObject FilterHeadersToObject(this System.Collections.Specialized.NameValueCollection self, bool isServerDocument)
        public static RavenJObject FilterHeadersToObject(this NameValueCollection self)
        {
            var metadata = new RavenJObject(StringComparer.OrdinalIgnoreCase);
            foreach (string header in self)
            {
                try
                {
                    if (PrefixesInHeadersToIgnoreClient.Any(prefix => header.StartsWith(prefix, StringComparison.OrdinalIgnoreCase)))
                        continue;
                    if (HeadersToIgnoreForClient.Contains(header))
                        continue;
                    var valuesNonDistinct = self.GetValues(header);
                    if (valuesNonDistinct == null)
                        continue;
                    var values = new HashSet<string>(valuesNonDistinct);
                    var headerName = CaptureHeaderName(header);
                    if (values.Count == 1)
                        metadata[headerName] = GetValueWithDates(values.First());
                    else
                        metadata[headerName] = new RavenJArray(values.Select(GetValueWithDates));
                }
                catch (Exception exc)
                {
                    throw new JsonReaderException(string.Concat("Unable to Filter Header: ", header), exc);
                }
            }
            return metadata;
        }

        [Obsolete("Use RavenFS instead.")]
        public static RavenJObject FilterHeadersAttachment(this IEnumerable<KeyValuePair<string, IEnumerable<string>>> self)
        {
            var filterHeaders = self.FilterHeadersToObject();
            string contentType = null;
            foreach (var keyValue in self)
            {
                if (keyValue.Key.Equals("Content-Type"))
                {
                    contentType = keyValue.Value.FirstOrDefault();
                    break;
                }
            }
            if (contentType != null)
                filterHeaders["Content-Type"] = contentType;

            return filterHeaders;
        }

        /// <summary>
        /// Filters the headers from unwanted headers
        /// </summary>
        /// <param name="self">HttpHeaders to filter</param>
        /// <param name="headersToIgnore">Headers to ignore</param>
        /// <param name="prefixesInHeadersToIgnore">Header prefixes to ignore</param>
        /// <returns></returns>
        public static RavenJObject FilterHeadersToObject(this IEnumerable<KeyValuePair<string, IEnumerable<string>>> self, HashSet<string> headersToIgnore, HashSet<string> prefixesInHeadersToIgnore)
        {
            var metadata = new RavenJObject(StringComparer.OrdinalIgnoreCase);
            foreach (var a in self)
            {
                var header = a.Key;
                try
                {
                    if (prefixesInHeadersToIgnore.Any(prefix => header.StartsWith(prefix, StringComparison.OrdinalIgnoreCase)))
                        continue;
                    if (headersToIgnore.Contains(header))
                        continue;
                    var valuesNonDistinct = a.Value;
                    if (valuesNonDistinct == null)
                        continue;
                    var values = new HashSet<string>(valuesNonDistinct);
                    var headerName = CaptureHeaderName(header);
                    if (values.Count == 1)
                        metadata[headerName] = GetValueWithDates(values.First());
                    else
                        metadata[headerName] = new RavenJArray(values.Select(GetValueWithDates));
                }
                catch (Exception exc)
                {
                    throw new JsonReaderException(string.Concat("Unable to Filter Header: ", header), exc);
                }
            }
            return metadata;
        }

        /// <summary>
        /// Filters the headers from unwanted headers
        /// </summary>
        /// <param name="self">The self.</param>
        /// <returns></returns>
        public static RavenJObject FilterHeadersToObject(this IEnumerable<KeyValuePair<string, IEnumerable<string>>> self)
        {
            return FilterHeadersToObject(self, HeadersToIgnoreClient, PrefixesInHeadersToIgnoreClient);
        }

        public static RavenJObject HeadersToObject(this HttpResponseMessage self)
        {
            var metadata = new RavenJObject(StringComparer.OrdinalIgnoreCase);

            var headers = self.Headers.Concat(self.Content.Headers);
            foreach (var a in headers)
            {
                var header = a.Key;

                try
                {
                    var headerName = CaptureHeaderName(header);

                    var valuesNonDistinct = a.Value;
                    if (valuesNonDistinct == null)
                        continue;

                    var values = new HashSet<string>(valuesNonDistinct);
                    if (values.Count == 1)
                    {
                        metadata[headerName] = GetValue(values.FirstOrDefault());
                    }
                    else
                    {
                        metadata[headerName] = new RavenJArray(values.Select(x => GetValue(x)));
                    }
                }
                catch (Exception exc)
                {
                    throw new JsonReaderException(string.Concat("Unable to build header: ", header), exc);
                }
            }

            return metadata;
        }

        private static string CaptureHeaderName(string header)
        {
            var lastWasDash = true;
            var sb = new StringBuilder(header.Length);

            foreach (var ch in header)
            {
                sb.Append(lastWasDash ? char.ToUpper(ch) : ch);

                lastWasDash = ch == '-';
            }

            return sb.ToString();
        }

        private static RavenJToken GetValue(string val)
        {
            try
            {
                if (val.StartsWith("{"))
                    return RavenJObject.Parse(val);
                if (val.StartsWith("["))
                    return RavenJArray.Parse(val);

                return new RavenJValue(Uri.UnescapeDataString(val));
            }
            catch (Exception exc)
            {
                throw new JsonReaderException(string.Concat("Unable to get value: ", val), exc);
            }
        }

        private static RavenJToken GetValueWithDates(string val)
        {
            try
            {
                if (val.StartsWith("{"))
                    return RavenJObject.Parse(val);
                if (val.StartsWith("["))
                    return RavenJArray.Parse(val);

                DateTime dateTime;
                if (DateTime.TryParseExact(val, Default.OnlyDateTimeFormat, CultureInfo.InvariantCulture, DateTimeStyles.RoundtripKind, out dateTime))
                {
                    if (val.EndsWith("Z"))
                        return DateTime.SpecifyKind(dateTime, DateTimeKind.Utc);
                    return new RavenJValue(dateTime);
                }

                DateTimeOffset dateTimeOffset;
                if (DateTimeOffset.TryParseExact(val, Default.DateTimeFormatsToRead, CultureInfo.InvariantCulture, DateTimeStyles.RoundtripKind, out dateTimeOffset))
                    return new RavenJValue(dateTimeOffset);

                return new RavenJValue(Uri.UnescapeDataString(val));
            }
            catch (Exception exc)
            {
                throw new JsonReaderException(string.Concat("Unable to get value: ", val), exc);
            }
        }
    }
}

File: Raven.Client.Java/src/main/java/net/ravendb/abstractions/extensions/MetadataExtensions.java
package net.ravendb.abstractions.extensions;

import java.text.ParseException;
import java.util.ArrayList;
import java.util.Date;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;
import java.util.Set;
import java.util.TreeSet;

import net.ravendb.abstractions.data.Constants;
import net.ravendb.abstractions.json.linq.RavenJArray;
import net.ravendb.abstractions.json.linq.RavenJObject;
import net.ravendb.abstractions.json.linq.RavenJToken;
import net.ravendb.abstractions.json.linq.RavenJValue;
import net.ravendb.abstractions.util.NetDateFormat;
import net.ravendb.client.utils.UrlUtils;


/**
 * Extensions for handling metadata
 */
public class MetadataExtensions {
  public final static Set<String> HEADERS_TO_IGNORE_CLIENT = new TreeSet<>(String.CASE_INSENSITIVE_ORDER);
  static {
    // Raven internal headers
    HEADERS_TO_IGNORE_CLIENT.add("Raven-Server-Build");
    HEADERS_TO_IGNORE_CLIENT.add("Raven-Client-Version");
    HEADERS_TO_IGNORE_CLIENT.add("Non-Authoritative-Information");
    HEADERS_TO_IGNORE_CLIENT.add("Raven-Timer-Request");
    HEADERS_TO_IGNORE_CLIENT.add("Raven-Authenticated-User");
    HEADERS_TO_IGNORE_CLIENT.add("Raven-Last-Modified");
    HEADERS_TO_IGNORE_CLIENT.add("Has-Api-Key");

    // COTS
    HEADERS_TO_IGNORE_CLIENT.add("Access-Control-Allow-Origin");
    HEADERS_TO_IGNORE_CLIENT.add("Access-Control-Max-Age");
    HEADERS_TO_IGNORE_CLIENT.add("Access-Control-Allow-Methods");
    HEADERS_TO_IGNORE_CLIENT.add("Access-Control-Request-Headers");
    HEADERS_TO_IGNORE_CLIENT.add("Access-Control-Allow-Headers");

    //proxy
    HEADERS_TO_IGNORE_CLIENT.add("Reverse-Via");
    HEADERS_TO_IGNORE_CLIENT.add("Persistent-Auth");
    HEADERS_TO_IGNORE_CLIENT.add("Allow");
    HEADERS_TO_IGNORE_CLIENT.add("Content-Disposition");
    HEADERS_TO_IGNORE_CLIENT.add("Content-Encoding");
    HEADERS_TO_IGNORE_CLIENT.add("Content-Language");
    HEADERS_TO_IGNORE_CLIENT.add("Content-Location");
    HEADERS_TO_IGNORE_CLIENT.add("Content-MD5");
    HEADERS_TO_IGNORE_CLIENT.add("Content-Range");
    HEADERS_TO_IGNORE_CLIENT.add("Content-Type");
    HEADERS_TO_IGNORE_CLIENT.add("Expires");
    // ignoring this header, we handle this internally
    HEADERS_TO_IGNORE_CLIENT.add("Last-Modified");
    // Ignoring this header, since it may
    // very well change due to things like encoding,
    // adding metadata, etc
    HEADERS_TO_IGNORE_CLIENT.add("Content-Length");
    // Special things to ignore
    HEADERS_TO_IGNORE_CLIENT.add("Keep-Alive");
    HEADERS_TO_IGNORE_CLIENT.add("X-Powered-By");
    HEADERS_TO_IGNORE_CLIENT.add("X-AspNet-Version");
    HEADERS_TO_IGNORE_CLIENT.add("X-Requested-With");
    HEADERS_TO_IGNORE_CLIENT.add("X-SourceFiles");
    // Request headers
    HEADERS_TO_IGNORE_CLIENT.add("Accept-Charset");
    HEADERS_TO_IGNORE_CLIENT.add("Accept-Encoding");
    HEADERS_TO_IGNORE_CLIENT.add("Accept");
    HEADERS_TO_IGNORE_CLIENT.add("Accept-Language");
    HEADERS_TO_IGNORE_CLIENT.add("Authorization");
    HEADERS_TO_IGNORE_CLIENT.add("Cookie");
    HEADERS_TO_IGNORE_CLIENT.add("Expect");
    HEADERS_TO_IGNORE_CLIENT.add("From");
    HEADERS_TO_IGNORE_CLIENT.add("Host");
    HEADERS_TO_IGNORE_CLIENT.add("If-Match");
    HEADERS_TO_IGNORE_CLIENT.add("If-Modified-Since");
    HEADERS_TO_IGNORE_CLIENT.add("If-None-Match");
    HEADERS_TO_IGNORE_CLIENT.add("If-Range");
    HEADERS_TO_IGNORE_CLIENT.add("If-Unmodified-Since");
    HEADERS_TO_IGNORE_CLIENT.add("Max-Forwards");
    HEADERS_TO_IGNORE_CLIENT.add("Referer");
    HEADERS_TO_IGNORE_CLIENT.add("TE");
    HEADERS_TO_IGNORE_CLIENT.add("User-Agent");
    //Response headers
    HEADERS_TO_IGNORE_CLIENT.add("Accept-Ranges");
    HEADERS_TO_IGNORE_CLIENT.add("Age");
    HEADERS_TO_IGNORE_CLIENT.add("Allow");
    HEADERS_TO_IGNORE_CLIENT.add("ETag");
    HEADERS_TO_IGNORE_CLIENT.add("Location");
    HEADERS_TO_IGNORE_CLIENT.add("Retry-After");
    HEADERS_TO_IGNORE_CLIENT.add("Server");
    HEADERS_TO_IGNORE_CLIENT.add("Set-Cookie2");
    HEADERS_TO_IGNORE_CLIENT.add("Set-Cookie");
    HEADERS_TO_IGNORE_CLIENT.add("Vary");
    HEADERS_TO_IGNORE_CLIENT.add("Www-Authenticate");
    // General
    HEADERS_TO_IGNORE_CLIENT.add("Cache-Control");
    HEADERS_TO_IGNORE_CLIENT.add("Connection");
    HEADERS_TO_IGNORE_CLIENT.add("Date");
    HEADERS_TO_IGNORE_CLIENT.add("Pragma");
    HEADERS_TO_IGNORE_CLIENT.add("Trailer");
    HEADERS_TO_IGNORE_CLIENT.add("Transfer-Encoding");
    HEADERS_TO_IGNORE_CLIENT.add("Upgrade");
    HEADERS_TO_IGNORE_CLIENT.add("Via");
    HEADERS_TO_IGNORE_CLIENT.add("Warning");

    // IIS Application Request Routing Module
    HEADERS_TO_IGNORE_CLIENT.add("X-ARR-LOG-ID");
    HEADERS_TO_IGNORE_CLIENT.add("X-ARR-SSL");
    HEADERS_TO_IGNORE_CLIENT.add("X-Forwarded-For");
    HEADERS_TO_IGNORE_CLIENT.add("X-Original-URL");
  }

  /**
   * Filters headers from unwanted headers
   * @param self
   * @return
   */
  public static RavenJObject filterHeaders(RavenJObject self) {
    if (self == null) {
      return null;
    }

    RavenJObject metadata = new RavenJObject(String.CASE_INSENSITIVE_ORDER);
    for (Entry<String, RavenJToken> header : self) {
      if (header.getKey().startsWith("Temp")) {
        continue;
      }
      if (header.getKey().equals(Constants.DOCUMENT_ID_FIELD_NAME)) {
        continue;
      }
      if (HEADERS_TO_IGNORE_CLIENT.contains(header.getKey())) {
        continue;
      }
      String headerName = captureHeaderName(header.getKey());
      metadata.add(headerName, header.getValue());
    }
    return metadata;
  }

  public static RavenJObject filterHeadersAttachment(Map<String, String> self) {
    RavenJObject filteredHeaders = filterHeaders(self);
    if (self.get("Content-Type") != null) {
      filteredHeaders.add("Content-Type", RavenJValue.fromObject(self.get("Content-Type")));
    }
    return filteredHeaders;
  }

  /**
   * Filters the headers from unwanted headers
   * @param headers
   * @return
   */
  public static RavenJObject filterHeaders(Map<String, String> headers) {
    RavenJObject metadata = new RavenJObject(String.CASE_INSENSITIVE_ORDER);
    for (Entry<String, String> header : headers.entrySet()) {
      if (header.getKey().startsWith("Temp")) {
        continue;
      }
      if (header.getKey().equals(Constants.DOCUMENT_ID_FIELD_NAME)) {
        continue;
      }
      if (HEADERS_TO_IGNORE_CLIENT.contains(header.getKey())) {
        continue;
      }
      Set<String> values = new HashSet<>();
      values.add(header.getValue());
      String headerName = captureHeaderName(header.getKey());
      if (values.size() == 1) {
        metadata.add(headerName, getValue(values.iterator().next()));
      } else {
        List<RavenJToken> headerValues = new ArrayList<>();
        for (String value : values) {
          headerValues.add(getValue(value));
        }
        metadata.add(headerName, new RavenJArray(headerValues.subList(0, Math.min(15, headerValues.size()))));
      }
    }
    return metadata;
  }

  private static String captureHeaderName(String header) {
    boolean lastWasDash = true;
    StringBuilder sb = new StringBuilder(header.length());

    for (int i = 0; i < header.length(); i++) {
      char ch = header.charAt(i);
      sb.append(lastWasDash ? Character.toUpperCase(ch) : ch);
      lastWasDash = ch == '-';

    }
    return sb.toString();
  }

  private static RavenJToken getValue(String val) {
    if (val.startsWith("{")) {
      return RavenJObject.parse(val);
    }
    if (val.startsWith("[")) {
      return RavenJArray.parse(val);
    }

    NetDateFormat netDateFormat = new NetDateFormat();
    try {
      Date date = netDateFormat.parse(val);
      return new RavenJValue(date);
    } catch (ParseException | IllegalArgumentException e) {
      //ignore
    }
    return new RavenJValue(UrlUtils.unescapeDataString(val));
  }

}
